lib/timers.js

```diff
@@ -10,36 +10,93 @@ const kOnTimeout = Timer.kOnTimeout | 0;
 // Timeout values > TIMEOUT_MAX are set to 1.
 const TIMEOUT_MAX = 2147483647; // 2^31-1
 
-// IDLE TIMEOUTS
+
+// HOW and WHY the timers implementation works the way it does.
+//
+// Since many parts of Node.js and user applications rely on timeouts, there may
+// be a very large amount of timeouts scheduled at any given time.
+// Therefore, it is very important that the timers implementation is very fast
+// and efficient.
+//
+// Note: It is suggested you first read though the lib/internal/linkedlist.js
+// linked list implementation. How it works is somewhat counter-intuitive to
+// most JavaScript code, and timers depend on it extensively.
+//
+// In order to be as efficient as possible, the implementation is entirely lazy.
+//
+// Object maps are kept which contain linked lists keyed by their duration in
+// milliseconds.
+// The linked lists within are TimerWrap C++ handles with a linked list appended
+// to their JavaScript representation.
+//
+// With this, constant-time insertion (append), removal, and timeout is possible
+// in the JavaScript layer. Any one list of timers is able to be sorted by just
+// appending to it because all timers share the same duration. Therefore, any
+// timer added later will always have been scheduled to timeouts later, thus
+// only needing to be appended. Any less-than constant overhead is left for a
+// lower and inherently faster layer to handle.
```

Comment:
Some questions:
* What is constant-time timeout?
* Do we support insertion? Or just append?
* Do we re-use the leading `TimerWrap` object across ticks? For example: if you `setTimeout(_=>setTimeout(_=>_, 50), 50)`, does the timer object get reused since a later tick placed a `Timeout` in the queue for 50 ms?